### PR TITLE
feat(soccer-stats): add event-based game timing system

### PR DESCRIPTION
**Pull request by JoA-MoS**
## Summary

Replace legacy game flow events and timing columns with a flexible event-based timing model that supports periods, stoppages, and metadata.

**Depends on:** #235 (TypeORM migrations infrastructure)

### New Event Types
| Event | Purpose |
|-------|---------|
| `GAME_START` / `GAME_END` | Game lifecycle |
| `PERIOD_START` / `PERIOD_END` | Period transitions (`metadata.period`: "1", "2", "OT1") |
| `STOPPAGE_START` / `STOPPAGE_END` | Clock stoppages (`metadata.reason`: "injury", etc.) |

### Event-Based Timing Model (Phase 1)

Timing values (`actualStart`, `firstHalfEnd`, `secondHalfStart`, `actualEnd`, `pausedAt`) are now derived from events instead of stored as columns:

- **`GameTimingService`**: Computes timing from timing events with batch loading support
- **`gameTimingLoader` DataLoader**: Solves N+1 query problem for timing field resolution
- **Event type caching**: Cached in `onModuleInit()` to avoid repeated lookups
- **`@ResolveField` resolvers**: Compute timing from events via DataLoader, fall back to legacy columns
- **Status change handling**: `games.service` creates timing events automatically when game status changes
- **Pause/Resume**: Creates STOPPAGE_START/STOPPAGE_END events instead of setting `pausedAt` column
- **Game reset**: Deletes timing events (not just clears columns)
- **Error handling**: Proper error propagation with throws instead of silent returns

### Backend
- Added `GameTimingService` with batch loading and event type caching
- Added `gameTimingLoader` DataLoader to solve N+1 query problem
- Added `@ResolveField` resolvers for all timing fields with legacy fallback
- Added virtual `period` field to `GameEvent` (GraphQL only, extracted from metadata)
- Refactored event types into shared `TIMING_EVENT_TYPES` constant
- Converted standalone migration script to TypeORM migration
- Updated `game-events.service` to use `GameTimingService` for duration calculations
- Added comprehensive test coverage (113 tests total)

### Frontend
- Added `isHalftimeEvent()` helper for backward compatibility
- Supports both legacy `HALFTIME` and new `PERIOD_END` with `period="1"`
- Added `period` field to GraphQL queries
- No changes required - timing fields still work via GraphQL resolvers

## Backward Compatibility

- Timing columns remain in database for backward compatibility during migration
- GraphQL resolvers compute timing from events first, fall back to column values
- Both old and new event types coexist in the database:
  - Legacy: `KICKOFF`, `HALFTIME`, `FULL_TIME`, `INJURY_STOPPAGE`
  - New: `GAME_START`, `GAME_END`, `PERIOD_START`, `PERIOD_END`, `STOPPAGE_START`, `STOPPAGE_END`

## Phase 2 (Future)

After verifying all games have timing events:
- Remove timing columns from Game entity
- Create migration to drop timing columns from database
- Add idempotency checks for timing event creation
- Add state validation for pause/resume events
- Extract shared TIMING_EVENT_TYPES constant

## Test Coverage

| File | Tests | Description |
|------|-------|-------------|
| `game-timing.service.spec.ts` | 25 | Timing computation, batching, caching, error handling |
| `games.service.spec.ts` | 12 | Status change events, pause/resume, reset |
| `event-types.service.spec.ts` | 8 | Event type seeding and validation |
| `startup.service.spec.ts` | 15 | Error classification and startup behavior |
| `game-summary.presentation.spec.ts` | 17 | isHalftimeEvent helper |

## Test plan

- [x] `pnpm nx lint soccer-stats-api` passes
- [x] `pnpm nx test soccer-stats-api` passes (113 tests)
- [x] `pnpm nx build soccer-stats-api` passes
- [x] `pnpm nx build soccer-stats-ui` passes
- [x] App startup creates new event types
- [ ] Manual test: Verify timing computed from events
- [ ] Manual test: Verify game status changes create timing events

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

**Review by JoA-MoS — COMMENTED** [review]
## 🔍 Code Review Results

**Score Threshold:** 80 | **Issues Found:** 1

---

### 🔴 Critical: `period` field missing from GraphQL schema

**File:** `apps/soccer-stats/api/src/entities/game-event.entity.ts:93-96`

**Issue:** The `period` getter is decorated with `@Field({ nullable: true })` but NestJS GraphQL cannot introspect the return type from a getter. This results in the field being **silently omitted from the schema**.

**Evidence:** The generated `schema.gql` shows `GameEvent` type without the `period` field (lines 198-225), even though the decorator is present in the entity.

**Current code:**
```typescript
@Field({ nullable: true })
get period(): string | undefined {
  return (this.metadata as { period?: string } | undefined)?.period;
}
```

**Fix:** Add explicit type to the `@Field` decorator:
```typescript
@Field(() => String, { nullable: true })
get period(): string | undefined {
  return (this.metadata as { period?: string } | undefined)?.period;
}
```

**Why this matters:** Without the explicit type, GraphQL queries for `period` on `GameEvent` will fail with a field-not-found error, breaking the timing events feature for clients.

---

*🤖 Generated by pr-review-toolkit*

File: apps/soccer-stats/api/src/modules/event-types/event-types.service.ts [hunk 1]
```diff
@@ -12,11 +12,69 @@ interface CreateEventTypeInput {
...
+    allowsParent: false,
+  },
+  {
+    name: 'STOPPAGE_END',
+    category: EventCategory.GAME_FLOW,
+    description: 'Clock resumes after stoppage',
+    requiresPosition: false,
+    allowsParent: false,
+  },
+];
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/api/src/modules/event-types/event-types.service.ts]: Addressed in commit 0ae885e. Created comprehensive test file `event-types.service.spec.ts` that covers:
1. TIMING_EVENT_TYPES constant structure (verifies all 6 timing event types are defined)
2. Correct properties for timing events (GAME_FLOW category, requiresPosition=false, allowsParent=false)
3. ensureNewEventTypesExist() behavior for creating new types and skipping existing ones
4. Idempotency testing for the method

File: apps/soccer-stats/api/src/scripts/migrate-timing-events.ts [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        await client.query(
+          `UPDATE game_events
+           SET "eventTypeId" = $1, metadata = $2
+           WHERE "eventTypeId" = $3`,
+          [
+            eventTypeMap.get(NEW_EVENT_TYPES.PERIOD_END),
+            JSON.stringify({ period: '1' }),
+            halftimeTypeId,
+          ],
+        );
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/api/src/scripts/migrate-timing-events.ts]: Addressed in commit 573adf3. Fixed the HALFTIME migration to preserve existing metadata using PostgreSQL JSONB merge:

```sql
SET metadata = (COALESCE(metadata::jsonb, '{}'::jsonb) || '{"period": "1"}'::jsonb)::json
```

This approach:
1. Handles null metadata gracefully with COALESCE
2. Merges the new period field into any existing metadata
3. Preserves all existing metadata properties while adding the period identifier

File: apps/soccer-stats/ui/src/app/components/presentation/game-summary.presentation.tsx [hunk 3]
```diff
@@ -98,6 +100,22 @@ type TimelineItem =
...
+  // Legacy event type
+  if (event.eventType?.name === 'HALFTIME') {
+    return true;
+  }
+  // New event type: PERIOD_END with period "1" marks end of first half
+  if (event.eventType?.name === 'PERIOD_END' && event.period === '1') {
+    return true;
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/ui/src/app/components/presentation/game-summary.presentation.tsx]: Addressed in commit 0ae885e. Created comprehensive test file `game-summary.presentation.spec.ts` with 17 tests covering:
1. Legacy HALFTIME event detection (with/without period metadata)
2. PERIOD_END with period='1' correctly identified as halftime
3. PERIOD_END with period='2', 'OT1', 'OT2' correctly NOT identified as halftime
4. PERIOD_END without period or with null period returns false
5. Other event types (GOAL, KICKOFF, GAME_START, GAME_END, PERIOD_START, SUBSTITUTION_IN) return false
6. Edge cases with null/undefined eventType

File: apps/soccer-stats/api/src/modules/games/game-timing.service.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+  }
+
+  /**
+   * Check if a game is currently paused.
+   */
+  async isGamePaused(gameId: string): Promise<boolean> {
+    const timing = await this.getGameTiming(gameId);
+    return timing.pausedAt !== undefined;
+  }
+}
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/api/src/modules/games/game-timing.service.ts]: Addressed in commit e33f469. Created comprehensive test file `game-timing.service.spec.ts` with 25 tests covering:

1. **onModuleInit caching** - Verifies event types are loaded and cached on startup
2. **getGameTiming** - Tests timing computation from various event combinations
3. **getGameTimingBatch** - Tests batch loading for multiple games (DataLoader integration)
4. **getGameDurationSeconds** - Tests duration calculation for different game states
5. **isGamePaused** - Tests pause state detection
6. **Stoppage handling** - Tests STOPPAGE_START/END pairing logic
7. **Error handling** - Tests behavior when event types are missing or database errors occur
8. **Edge cases** - Tests games with no events and incomplete event sequences

File: apps/soccer-stats/api/src/modules/games/game-timing.service.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    // Get timing event types
+    const timingEventTypes = await this.eventTypeRepository.find({
+      where: [
+        { name: 'GAME_START' },
+        { name: 'GAME_END' },
+        { name: 'PERIOD_START' },
+        { name: 'PERIOD_END' },
+        { name: 'STOPPAGE_START' },
+        { name: 'STOPPAGE_END' },
+      ],
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/api/src/modules/games/game-timing.service.ts]: Acknowledged as a valid improvement. The TIMING_EVENT_TYPES constant is currently defined in two places:
- `game-timing.service.ts` (for computing timing)
- `games.service.ts` (for creating/deleting timing events)

**Trade-offs considered:**
- Creating a shared constant would require careful module dependency management to avoid circular dependencies
- The current duplication is limited to 6 event type names that are unlikely to change

**Recommendation:** Track this as a low-priority refactoring task for Phase 2, when the timing column migration is complete and we can consolidate timing-related code. For now, the implementation is stable and tested.

File: apps/soccer-stats/api/src/modules/games/games.service.ts [hunk 6]
```diff
@@ -158,17 +158,42 @@ export class GamesService {
...
+        // Clear only timing events (event-based timing model)
+        const timingEventTypes = await this.eventTypeRepository.find({
+          where: [
+            { name: 'GAME_START' },
+            { name: 'GAME_END' },
+            { name: 'PERIOD_START' },
+            { name: 'PERIOD_END' },
+            { name: 'STOPPAGE_START' },
+            { name: 'STOPPAGE_END' },
+          ],
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/api/src/modules/games/games.service.ts]: Same as the comment on game-timing.service.ts - acknowledged as a valid improvement for Phase 2. The duplication is intentional for now to avoid circular module dependencies, and the 6 event type names are stable reference data.

File: apps/soccer-stats/api/src/modules/games/games.service.ts [hunk 7]
```diff
@@ -363,4 +402,99 @@ export class GamesService {
...
+    }
+
+    const event = this.gameEventRepository.create({
+      gameId,
+      eventTypeId: eventType.id,
+      gameMinute: 0,
+      gameSecond: 0,
+    });
+    await this.gameEventRepository.save(event);
+  }
```
[the remark this answers is not in the document]
Comment (JoA-MoS) [reply apps/soccer-stats/api/src/modules/games/games.service.ts]: Valid concern acknowledged for Phase 2. The current implementation creates events whenever `pausedAt` changes.

**Current behavior:**
- `pausedAt = Date` → creates STOPPAGE_START
- `pausedAt = null` → creates STOPPAGE_END
- Unmatched STOPPAGE_START events mean the game is paused

**Phase 2 improvements to consider:**
1. Query current pause state before creating events
2. Only create STOPPAGE_START if not already paused
3. Only create STOPPAGE_END if currently paused (has unmatched STOPPAGE_START)

**Mitigation:** The `computeTimingFromEvents()` algorithm handles unbalanced sequences gracefully - it tracks the last STOPPAGE_START and clears it on STOPPAGE_END. Multiple consecutive STOPPAGE_START events just update the pause timestamp; multiple STOPPAGE_END events are no-ops.
